Applications/Spire/Include/Spire/Ui/AssociativeValueModel.hpp

```diff
@@ -1,38 +1,47 @@
 #ifndef SPIRE_ASSOCIATIVE_VALUE_MODEL_HPP
 #define SPIRE_ASSOCIATIVE_VALUE_MODEL_HPP
 #include <functional>
+#include <boost/functional/hash.hpp>
 #include <boost/signals2/shared_connection_block.hpp>
 #include "Spire/Ui/Ui.hpp"
 #include "Spire/Ui/LocalValueModel.hpp"
 
-namespace Spire {
-namespace Details {
-  struct AssociatedBooleanModel {
-    std::shared_ptr<BooleanModel> m_model;
-    boost::signals2::scoped_connection m_connection;
-  };
-
-  template <typename T, typename U = boost::optional<T>>
-  struct OptionalHash {
-    std::size_t operator()(const U& value) const;
+namespace std {
+  template<typename T>
+  struct hash<boost::optional<T>> {
+    size_t operator ()(const boost::optional<T>& value) const {
+      if(!value) {
+        return -3333;
+      }
+      return boost::hash_value(value.get());
+    }
   };
 }
 
+namespace Spire {
+
   /** 
    * Associates BooleanModels with corresponding values, and ensures that
    * only a single model has a value of true at one time.
    */
-  template<typename T, typename U = std::hash<T>>
+  template<typename T>
   class AssociativeValueModel : public ValueModel<T> {
     public:
 
       using Type = T;
 
-      using Hash = U;
-
       /** Constructs an AssociativeValueModel. */
       AssociativeValueModel();
 
+      /**
+      * Constructs an AssociativeValueModel with a default value. The default
+      * value is the value that will become current if all associated models
+      * have a value of false. If the provided value is none, there is no
+      * default value.
+      * @param default_value The default value.
+      */
+      AssociativeValueModel(const boost::optional<T>& default_value);
```

Comment:
Why is this ```boost::optional<T>``` instead of ```T```?

Reply:
Don't leak implementation details into the API. The API must be designed for what's best for the user, not the implementor.